### PR TITLE
Add support for collocation of transformer and predictor

### LINKED ISSUE #2349: collocate of transformer and triton predictor throwing error
/kind bug

**What steps did you take and what happened:**
I am working with transformer and onnx runtime so have to use triton server. My application is very latency specific, but due to two pods I believe there is a latency overhead as given in this issue: https://github.com/kserve/kserve/issues/1957
So, I want to achieve collocate feature so that in one pod both transformer and predictor will run. But not able to achive this and throwing error in predictor. I tested with single predictor only and its working, but with transformer throwing error, the manifest is given below:

```
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "kserve-inference-service"
spec:
  predictor:
    serviceAccountName: sa
    containers:
    - image: demoregistry.azurecr.io/transformer:v1
      name: transformer-container
      command:
      - "python"
      - "-m"
      - "transformer"
      args:
        - --user_entity_ids
        - id
        - --user_feature_refs
        - user_feature_view:pFeatures
        - --product_entity_ids
        - product_id
        - --product_feature_refs
        - product_view:Product_Features
        - --predictor_host
        - localhost:8080
      ports:
        - containerPort: 8000
          protocol: TPS
      env:
        - name: BLOB_KEY
          value: "key-value"
    - image: nvcr.io/nvidia/tritonserver:20.10-py3
      name: triton-server
      args:
        - tritonserver
        - --model-store=/mnt/models
        - --grpc-port=9000
        - --http-port=8080
        - --allow-grpc=true
        - --allow-http=true
      env:
        - name: STORAGE_URI
          value: https://blobdl.blob.core.windows.net/model/onnx/
```

**What did you expect to happen:**
I was aiming that both transformer and predictor would run in one single pod, don't know where is the issue here, or this collocate can only supported with pytorch not triton server.

**Environment:**

- Istio Version: v1.10.3
- Knative Version: v1.0.0
- KFServing Version: v0.8.0
- Kubeflow version: X
- Kfdef:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube version: X
- Kubernetes version: (use `kubectl version`): 1.21.9
- OS (e.g. from `/etc/os-release`):  ubuntu

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2349 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Add support for collocation of transformer and predictor in the same pod
```

### CODE REVIEW

File: docs/samples/v1beta1/transformer/collocation/collocation.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+metadata:
+  name: custom-transformer-collocation
+spec:
+  predictor:
+    containers:
+      - name: kserve-container
+        image: kserve/custom-model-grpc:latest
+        args:
+          - --model_name=custom-model
+          - --protocol=grpc-v2
```
Comment (yuzisun) [inline docs/samples/v1beta1/transformer/collocation/collocation.yaml]: We do not need this argument for predictor, it is only for specifying the protocol between transformer and predictor

File: docs/samples/v1beta1/transformer/collocation/collocation.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+          - --grpc_port=8082
```
Comment (yuzisun) [inline docs/samples/v1beta1/transformer/collocation/collocation.yaml:12]: for gRPC I think we need to expose the port with
```yaml
ports:
   - containerPort: 8082
     name: h2c
     protocol: TCP
```

Comment (sivanantha321) [reply docs/samples/v1beta1/transformer/collocation/collocation.yaml:12]: Knative does not allow setting ports for more than one container. So predictor cannot be exposed.

File: docs/samples/v1beta1/transformer/collocation/collocation.yaml (hunk 1 above)
Comment (sivanantha321) [reply docs/samples/v1beta1/transformer/collocation/collocation.yaml]: done

File: docs/samples/v1beta1/transformer/collocation/collocation.yaml (hunk 2 above)
Comment (yuzisun) [reply docs/samples/v1beta1/transformer/collocation/collocation.yaml:12]: Actually you do not need to expose the port as it is local networking call

**Comment by yuzisun** [comment]
Awesome work! Thanks @sivanantha321 !

/lgtm
/approve

File: docs/samples/v1beta1/transformer/collocation/collocation.yaml (hunk 2 above)
Comment (sivanantha321) [reply docs/samples/v1beta1/transformer/collocation/collocation.yaml:12]: Exactly!

**Pull request merged**